.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,639 @@
+          file: Dockerfile_s3-deploy-env
+          push: false
+          build-args: |
+            DEPLOYER_UID=${{ steps.s3-lib.outputs.deployer_uid }}
+            DEPLOYER_GID=${{ steps.s3-lib.outputs.deployer_gid }}
+
+      - name: Deploy to S3
+        id: s3-deploy
+        env:
+          BASE_DIR: /host/images
+          S3_CMD: "s4cmd --access-key=${{ secrets.S3_ACCESS_KEY }} --secret-key=${{ secrets.S3_SECRET_KEY }} --API-ServerSideEncryption=AES256"
+          S3_SYNC_OPTS: "--recursive --API-ACL=${{ steps.s3-lib.outputs.s3_policy }}"
+          S3_BUCKET: "${{ inputs.s3-bucket }}"
+          SLUG: "${{ needs.build.outputs.device_slug }}"
+          DEPLOY_ARTIFACT: "${{ needs.build.outputs.deploy_artifact }}"
+          HOSTOS_VERSION: "${{ needs.build.outputs.os_version }}"
+          # TODO: images path must be found in artifacts
+          IMAGES_PATH: ${{ github.workspace }}/build-artifacts
+        run: |
+          docker run --rm -t --user deployer \
+            -e BASE_DIR \
+            -e S3_CMD \
+            -e S3_SYNC_OPTS \
+            -e S3_BUCKET \
+            -e SLUG \
+            -e DEPLOY_ARTIFACT \
+            -e HOSTOS_VERSION \
+            -v "${IMAGES_PATH}:/host/images" \
+            "${{ steps.helper-image.outputs.imageid }}" /bin/sh -e -c '\
+              echo "${HOSTOS_VERSION}" > "/host/images/${SLUG}/latest"
+              if [ "$DEPLOY_ARTIFACT" != "docker-image" ]; then
+                /usr/src/app/node_modules/.bin/ts-node /usr/src/app/scripts/prepare.ts
+              fi
+              if [ -z "$($S3_CMD ls s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/)" ] || [ -n "$($S3_CMD ls s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/IGNORE)" ]; then
+                touch /host/images/${SLUG}/${HOSTOS_VERSION}/IGNORE
+                $S3_CMD del -rf s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}
+                $S3_CMD put /host/images/${SLUG}/${HOSTOS_VERSION}/IGNORE s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/
+                $S3_CMD $S3_SYNC_OPTS dsync /host/images/${SLUG}/${HOSTOS_VERSION}/ s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/
+                $S3_CMD put /host/images/${SLUG}/latest s3://${S3_BUCKET}/${SLUG}/ --API-ACL=public-read -f
+                $S3_CMD del s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/IGNORE
+              else
+                echo "WARNING: Deployment already done for ${SLUG} at version ${HOSTOS_VERSION}"
+              fi
+              '
+
+  ami-deploy:
+    name: AMI Deploy
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    needs: build
+    if: inputs.deploy-ami == true
+
+    env:
+      MACHINE: "${{ inputs.machine }}"
+      BALENARC_BALENA_URL: "${{ inputs.balena-environment }}"
+      WORKSPACE: ${{ github.workspace }}
+      VERBOSE: verbose
+
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+
+    steps:
+      # TODO: clone balena-yocto-scripts @ yocto_scripts_ref instead
```

Comment:
Same concersn about re-cloning - we should opmitize.

Reply:
See above answer, we are using 4 separate jobs to make use of parallelism and each job is a fresh workspace.

If this turns out to not be faster, we can revert to one job and workspace.